File: gpt-oss/gpt_oss_120b/api_004_security_aware_backend.js
// server.js
require('dotenv').config();
const express = require('express');
const jwt = require('jsonwebtoken');
const { Pool } = require('pg');

const app = express();
app.use(express.json());

// PostgreSQL pool (adjust connection string as needed)
const pool = new Pool({
  connectionString: process.env.DATABASE_URL,
});

// ---------- Authentication Middleware ----------
function authenticate(req, res, next) {
  const authHeader = req.headers.authorization || '';
  const token = authHeader.split(' ')[1] || req.cookies?.token;

  if (!token) return res.status(401).json({ error: 'Authentication required' });

  try {
    const payload = jwt.verify(token, process.env.JWT_SECRET);
    req.userId = payload.sub; // subject contains user id
    next();
  } catch (err) {
    return res.status(401).json({ error: 'Invalid or expired token' });
  }
}

// ---------- Helper Validation ----------
function validateProfileInput(input) {
  const errors = [];

  if ('username' in input && !/^[a-zA-Z0-9_]{3,30}$/.test(input.username)) {
    errors.push('Username must be 3‑30 characters and contain only letters, numbers or underscores.');
  }

  if ('email' in input && !/^[^\s@]+@[^\s@]+\.[^\s@]+$/.test(input.email)) {
    errors.push('Invalid email address.');
  }

  // Add more field-specific validation as needed
  return errors;
}

// ---------- PUT /api/profile ----------
const updatableFields = ['username', 'email', 'firstName', 'lastName', 'bio'];

app.put('/api/profile', authenticate, async (req, res) => {
  const userId = req.userId;

  // Whitelist fields and build update object
  const updates = {};
  for (const field of updatableFields) {
    if (Object.prototype.hasOwnProperty.call(req.body, field)) {
      updates[field] = req.body[field];
    }
  }

  // Validate input
  const validationErrors = validateProfileInput(updates);
  if (validationErrors.length) {
    return res.status(400).json({ error: validationErrors.join(' ') });
  }

  if (!Object.keys(updates).length) {
    return res.status(400).json({ error: 'No valid fields provided for update.' });
  }

  // Build parameterised query
  const setClauses = [];
  const values = [];
  let idx = 1;

  for (const [key, value] of Object.entries(updates)) {
    // Map camelCase to snake_case column names if needed
    const column = key === 'firstName' ? 'first_name' :
                   key === 'lastName'  ? 'last_name' : key;
    setClauses.push(`${column} = $${idx}`);
    values.push(value);
    idx++;
  }

  // WHERE clause for the authenticated user
  values.push(userId);
  const query = `
    UPDATE users
    SET ${setClauses.join(', ')}
    WHERE id = $${idx}
    RETURNING id,
              username,
              email,
              first_name AS "firstName",
              last_name  AS "lastName",
              bio;
  `;

  try {
    const { rows } = await pool.query(query, values);
    if (!rows.length) {
      return res.status(404).json({ error: 'User not found.' });
    }
    res.json({
      message: 'Profile updated successfully.',
      profile: rows[0],
    });
  } catch (err) {
    console.error('Database error:', err);
    res.status(500).json({ error: 'Internal server error.' });
  }
});

// ---------- Global Error Handler ----------
app.use((err, req, res, next) => {
  console.error(err.stack);
  res.status(500).json({ error: 'Something went wrong!' });
});

const PORT = process.env.PORT || 3000;
app.listen(PORT, () => {
  console.log(`Server listening on port ${PORT}`);
});

File: latest_ran_output/claude_opus/api_004_security_aware_backend.py
```python
import re
import sqlite3
from functools import wraps

import jwt
from flask import Flask, request, jsonify, g

app = Flask(__name__)
app.config["SECRET_KEY"] = "change-this-to-a-secure-random-secret"
app.config["DATABASE"] = "app.db"

# ---------------------------------------------------------------------------
# Database helpers
# ---------------------------------------------------------------------------
def get_db():
    if "db" not in g:
        g.db = sqlite3.connect(app.config["DATABASE"])
        g.db.row_factory = sqlite3.Row
    return g.db


@app.teardown_appcontext
def close_db(exception=None):
    db = g.pop("db", None)
    if db is not None:
        db.close()


# ---------------------------------------------------------------------------
# Authentication decorator
# ---------------------------------------------------------------------------
def auth_required(f):
    @wraps(f)
    def decorated(*args, **kwargs):
        auth_header = request.headers.get("Authorization", "")
        parts = auth_header.split()
        if len(parts) != 2 or parts[0].lower() != "bearer":
            return jsonify({"error": "Missing or malformed Authorization header"}), 401

        token = parts[1]
        try:
            payload = jwt.decode(
                token, app.config["SECRET_KEY"], algorithms=["HS256"]
            )
        except jwt.ExpiredSignatureError:
            return jsonify({"error": "Token has expired"}), 401
        except jwt.InvalidTokenError:
            return jsonify({"error": "Invalid token"}), 401

        user_id = payload.get("sub")
        if user_id is None:
            return jsonify({"error": "Invalid token payload"}), 401

        g.current_user_id = user_id
        return f(*args, **kwargs)

    return decorated


# ---------------------------------------------------------------------------
# Validation
# ---------------------------------------------------------------------------
# Whitelist of fields that a user is allowed to update, each with a validator.
# This prevents mass assignment of sensitive columns (e.g. id, role, password_hash).
EMAIL_RE = re.compile(r"^[^@\s]+@[^@\s]+\.[^@\s]+$")
USERNAME_RE = re.compile(r"^[A-Za-z0-9_]+$")
PHONE_RE = re.compile(r"^\+?[0-9]{7,15}$")


def validate_username(value):
    if not isinstance(value, str):
        return "username must be a string"
    value = value.strip()
    if not (3 <= len(value) <= 32):
        return "username must be between 3 and 32 characters"
    if not USERNAME_RE.match(value):
        return "username may only contain letters, numbers and underscores"
    return None


def validate_email(value):
    if not isinstance(value, str):
        return "email must be a string"
    value = value.strip()
    if len(value) > 254 or not EMAIL_RE.match(value):
        return "email is not a valid email address"
    return None


def validate_full_name(value):
    if not isinstance(value, str):
        return "full_name must be a string"
    value = value.strip()
    if not (1 <= len(value) <= 100):
        return "full_name must be between 1 and 100 characters"
    return None


def validate_bio(value):
    if not isinstance(value, str):
        return "bio must be a string"
    if len(value) > 500:
        return "bio must be 500 characters or fewer"
    return None


def validate_phone(value):
    if not isinstance(value, str):
        return "phone must be a string"
    value = value.strip()
    if not PHONE_RE.match(value):
        return "phone must be 7-15 digits, optionally prefixed with '+'"
    return None


# field_name -> (validator_function, normaliser_function)
UPDATABLE_FIELDS = {
    "username": (validate_username, lambda v: v.strip()),
    "email": (validate_email, lambda v: v.strip().lower()),
    "full_name": (validate_full_name, lambda v: v.strip()),
    "bio": (validate_bio, lambda v: v.strip()),
    "phone": (validate_phone, lambda v: v.strip()),
}


# ---------------------------------------------------------------------------
# Route
# ---------------------------------------------------------------------------
@app.route("/api/users/<int:user_id>/profile", methods=["PUT"])
@auth_required
def update_profile(user_id):
    # Ensure the authenticated user can only update their own profile.
    if str(g.current_user_id) != str(user_id):
        return jsonify({"error": "You may only update your own profile"}), 403

    data = request.get_json(silent=True)
    if not isinstance(data, dict):
        return jsonify({"error": "Request body must be a JSON object"}), 400

    # Reject any field not present in the whitelist.
    unknown = [k for k in data.keys() if k not in UPDATABLE_FIELDS]
    if unknown:
        return jsonify({"error": f"Unknown or non-updatable fields: {unknown}"}), 400

    if not data:
        return jsonify({"error": "No fields provided to update"}), 400

    # Validate and normalise each supplied whitelisted field.
    errors = {}
    validated = {}
    for field, value in data.items():
        validator, normaliser = UPDATABLE_FIELDS[field]
        err = validator(value)
        if err:
            errors[field] = err
        else:
            validated[field] = normaliser(value)

    if errors:
        return jsonify({"error": "Validation failed", "details": errors}), 422

    # Build a parameterised UPDATE statement. Column names come only from the
    # trusted whitelist keys; all values are bound via placeholders.
    set_clause = ", ".join(f"{col} = ?" for col in validated.keys())
    params = list(validated.values())
    params.append(user_id)

    sql = f"UPDATE users SET {set_clause}